In [84]:
import pandas as pd # A general purpose Python library for data analysis
import numpy as np # A library for scientific computing in Python (e.g., provides high-performance multi-dimensional array objects and operations)

import matplotlib.pyplot as plt # a plotting library for Python and NumPy (readily customizable)
import seaborn as sns # Another plotting library for Python (fewer syntax, excellent default themes, behind the scenes, it uses matplotlib)
import time

## Knowledge Streams 2024

In this notebook, we will learn about the key data structures provided by the Pandas library: **Data Frames, Series, and Indices**.

In addition, we will learn about the following operations:
* How to access data contained in these structures?
* How to read files (e.g., csv, xlsx, sql) to create these structures?
* How to carry out different data manipulation tasks using these structures?

`Dataset`: US elections with information about candidates, their party, votes won, year of election and the result.

## Reading in Data Frames from Files
We'll be using **read_csv** today. Note that this file reading function does all the *data parsing* for you, which is very useful.

Before loading a file into a dataframe, let's first take a look at the **elections.csv** file

In [85]:
#Load csv file and print shape
# Code here
elections=pd.read_csv('./data/elections.csv')

# how many observation and features are given
#Code here
print(f'.\nThe data set has {elections.shape[0]} observations and {elections.shape[1]} features')

.
The data set has 182 observations and 6 features


In [86]:
# We can use the **head command** to show only a few rows of a dataframe from start.
# Code here
print(elections.head())
#Use **tail command** to show last few observation.
# code here
print(elections.tail())

   Year          Candidate                  Party  Popular vote Result  \
0  1824     Andrew Jackson  Democratic-Republican        151271   loss   
1  1824  John Quincy Adams  Democratic-Republican        113142    win   
2  1828     Andrew Jackson             Democratic        642806    win   
3  1828  John Quincy Adams    National Republican        500897   loss   
4  1832     Andrew Jackson             Democratic        702735    win   

           %  
0  57.210122  
1  42.789878  
2  56.203927  
3  43.796073  
4  54.574789  
     Year       Candidate        Party  Popular vote Result          %
177  2016      Jill Stein        Green       1457226   loss   1.073699
178  2020    Joseph Biden   Democratic      81268924    win  51.311515
179  2020    Donald Trump   Republican      74216154   loss  46.858542
180  2020    Jo Jorgensen  Libertarian       1865724   loss   1.177979
181  2020  Howard Hawkins        Green        405035   loss   0.255731


In [87]:
# #The `read_csv` command lets us specify a **column to use an index**. For example, we could have used __Year__ as the index.
# #Code here
# elections=pd.read_csv('./data/elections.csv',index_col='Year')
# elections.head()

In [88]:
#Alternately, we could have used the **set_index** commmand on the dataframe to set a particular column as index.
# code here
# elections.set_index('Party',inplace=True)
# elections.head()

# Caution:
The **set_index command** (along with all other data frame methods) **does not modify the dataframe**, i.e., the original "elections" is untouched. Note: There is a flag called "inplace" which does modify the calling dataframe (e.g., `elections.set_index("Party",inplace=True)`).

## Duplicate Columns?
By contast, column names MUST be unique. For example, if we try to read in a file for which column names are not unique, Pandas will automatically any duplicates. Load duplicate_columns.csv

In [89]:
#Answer Here
duplicate=pd.read_csv('./data/duplicate_columns.csv')
duplicate

name    name.1      flavor
0   john     smith     vanilla
1  zhang      shan   chocolate
2  fulan  alfulani  strawberry
3   hong   gildong      banana

## The [ ] Operator & Indexing

The DataFrame class has an indexing operator **[ ]** (also known as the 'brack' operator) that lets you do a variety of different things. If your provide a String to the **[ ]** operator, you get back a ***Series*** corresponding to the requested label.

1.Use **[ ]** to display different columns

2.Use List retrive multiple columns

In [90]:
# Display and Retrieve multiple columns from the election data frame, the resultant would be the list for every column.
#Code here
candidate_list = elections['Candidate'].tolist()
popular_vote_list = elections['Popular vote'].tolist()

# Display the lists
print("Candidate List:", candidate_list)
print("\nPopular Vote List:", popular_vote_list)

Candidate List: ['Andrew Jackson', 'John Quincy Adams', 'Andrew Jackson', 'John Quincy Adams', 'Andrew Jackson', 'Henry Clay', 'William Wirt', 'Hugh Lawson White', 'Martin Van Buren', 'William Henry Harrison', 'Martin Van Buren', 'William Henry Harrison', 'Henry Clay', 'James Polk', 'Lewis Cass', 'Martin Van Buren', 'Zachary Taylor', 'Franklin Pierce', 'John P. Hale', 'Winfield Scott', 'James Buchanan', 'John C. Frémont', 'Millard Fillmore', 'Abraham Lincoln', 'John Bell', 'John C. Breckinridge', 'Stephen A. Douglas', 'Abraham Lincoln', 'George B. McClellan', 'Horatio Seymour', 'Ulysses Grant', 'Horace Greeley', 'Ulysses Grant', 'Rutherford Hayes', 'Samuel J. Tilden', 'James B. Weaver', 'James Garfield', 'Winfield Scott Hancock', 'Benjamin Butler', 'Grover Cleveland', 'James G. Blaine', 'John St. John', 'Alson Streeter', 'Benjamin Harrison', 'Clinton B. Fisk', 'Grover Cleveland', 'Benjamin Harrison', 'Grover Cleveland', 'James B. Weaver', 'John Bidwell', 'John M. Palmer', 'Joshua Lever

The **[ ]** operator also accepts a list of strings. In this case, you get back a **DataFrame** corresponding to the requested strings.

In [91]:
# Answer Here
list=elections[['Candidate','Popular vote']]
list.head()

Candidate  Popular vote
0     Andrew Jackson        151271
1  John Quincy Adams        113142
2     Andrew Jackson        642806
3  John Quincy Adams        500897
4     Andrew Jackson        702735

A list of one label also returns a DataFrame. This can be handy if you want your results as a DataFrame, not a series.

Note that we can also use the **to_frame** method to turn a Series into a DataFrame.

Extract one col name "Candidates" from DataFrame it will be a series. Convert series into a DataFrame.

In [92]:
# Answer Here
candidate_series = elections['Candidate']
candidate_series.head()

0       Andrew Jackson
1    John Quincy Adams
2       Andrew Jackson
3    John Quincy Adams
4       Andrew Jackson
Name: Candidate, dtype: object

In [93]:
# Answer Here
candidate_df = candidate_series.to_frame()
candidate_df.head()

Candidate
0     Andrew Jackson
1  John Quincy Adams
2     Andrew Jackson
3  John Quincy Adams
4     Andrew Jackson

### Row Indexing

The `[]` operator also accepts numerical slices as arguments. In this case, we are indexing by row, not column!

Extract few rows from DataFrame

In [94]:
# Answer Here
elections.iloc[0:5]

Year          Candidate                  Party  Popular vote Result  \
0  1824     Andrew Jackson  Democratic-Republican        151271   loss   
1  1824  John Quincy Adams  Democratic-Republican        113142    win   
2  1828     Andrew Jackson             Democratic        642806    win   
3  1828  John Quincy Adams    National Republican        500897   loss   
4  1832     Andrew Jackson             Democratic        702735    win   

           %  
0  57.210122  
1  42.789878  
2  56.203927  
3  43.796073  
4  54.574789

If you provide a single argument to the `[]` operator, it tries to use it as a name. This is true even if the argument passed to **[ ]** is an integer.

In [95]:
#elections[0] #this does not work, try uncommenting this to see it fail in action, woo
elections.iloc[0]

Year                             1824
Candidate              Andrew Jackson
Party           Democratic-Republican
Popular vote                   151271
Result                           loss
%                           57.210122
Name: 0, dtype: object

The following cells allow you to **test your understanding**. Let's go over the summary of what we have learnt (see slides).

# Creating DataFrames
Create DataFrame using List and Columns name.

In [96]:
# Answer Here
# Lists for the DataFrame
votes_list = [151271, 113142, 642806]
candidates_list = ['Andrew Jackson', 'John Quincy Adams', 'Andrew Jackson']

# Create DataFrame directly
df_list = pd.DataFrame({
    'Name': candidates_list,
    'Votes': votes_list,
    
})
df_list

Name   Votes
0     Andrew Jackson  151271
1  John Quincy Adams  113142
2     Andrew Jackson  642806

Creating DataFrames using **Dictionary**.

In [97]:
# Answer Here
data = {
    'Name': ['John', 'Emily', 'Anna'],
    'Age': [25, 32, 29],
    'City': ['New York', 'London', 'Paris']
}

# Create DataFrame from the dictionary
df_dict = pd.DataFrame(data)
df_dict

Name  Age      City
0   John   25  New York
1  Emily   32    London
2   Anna   29     Paris

## Filtering via Boolean Array Selection

The `[]` operator also supports array of booleans as an input. In this case, the array must be exactly as long as the number of rows. The result is a **filtered version of the data frame**, where **only rows corresponding to True appear**.

In [98]:
elections['Result_bool'] = elections['Result'].replace({'win': True, 'loss': False})
elections.head()

C:\Users\nasrullah\AppData\Local\Temp\ipykernel_15144\3000792151.py:1: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  elections['Result_bool'] = elections['Result'].replace({'win': True, 'loss': False})


Year          Candidate                  Party  Popular vote Result  \
0  1824     Andrew Jackson  Democratic-Republican        151271   loss   
1  1824  John Quincy Adams  Democratic-Republican        113142    win   
2  1828     Andrew Jackson             Democratic        642806    win   
3  1828  John Quincy Adams    National Republican        500897   loss   
4  1832     Andrew Jackson             Democratic        702735    win   

           %  Result_bool  
0  57.210122        False  
1  42.789878         True  
2  56.203927         True  
3  43.796073        False  
4  54.574789         True

One very common task in Data Science is **filtering**. Boolean Array Selection is one way to achieve this in Pandas. We start by observing that **logical operators** like the equality operator can be applied to **Pandas Series data** to generate a **Boolean Array**.

Compare the 'Result' column to the String 'win' and Show results

In [99]:
#Answer Here
mask_result=elections['Result'] == 'win'
winning_candidates=elections[mask_result]
winning_candidates.head()

Year               Candidate                  Party  Popular vote Result  \
1   1824       John Quincy Adams  Democratic-Republican        113142    win   
2   1828          Andrew Jackson             Democratic        642806    win   
4   1832          Andrew Jackson             Democratic        702735    win   
8   1836        Martin Van Buren             Democratic        763291    win   
11  1840  William Henry Harrison                   Whig       1275583    win   

            %  Result_bool  
1   42.789878         True  
2   56.203927         True  
4   54.574789         True  
8   52.272472         True  
11  53.051213         True

Compare the 'Party' column to the String 'Democratic' and Show results

In [100]:
#Answer Here
mask_democratic = elections['Party'] == 'Democratic'
democratic = elections[mask_democratic]
democratic.head()


Year         Candidate       Party  Popular vote Result          %  \
2   1828    Andrew Jackson  Democratic        642806    win  56.203927   
4   1832    Andrew Jackson  Democratic        702735    win  54.574789   
8   1836  Martin Van Buren  Democratic        763291    win  52.272472   
10  1840  Martin Van Buren  Democratic       1128854   loss  46.948787   
13  1844        James Polk  Democratic       1339570    win  50.749477   

    Result_bool  
2          True  
4          True  
8          True  
10        False  
13         True

The output of the logical operator applied to the Series is **another Series with the same name and index, but of datatype boolean**.

These boolean Series can be used as an argument to the `[]` operator.

Creates  DataFrame of all election winners since 1980.

In [130]:
#Answer Here
# Filtering for winners since 1980
winners_since_1980 = elections[(elections['Year'] >= 1980) & (elections['Result'] == 'win')]
# Displaying the result
iswin=pd.DataFrame(winners_since_1980)
iswin


Year          Candidate       Party  Popular vote Result          %  \
131  1980      Ronald Reagan  Republican      43903230    win  50.897944   
133  1984      Ronald Reagan  Republican      54455472    win  59.023326   
135  1988  George H. W. Bush  Republican      48886597    win  53.518845   
140  1992       Bill Clinton  Democratic      44909806    win  43.118485   
144  1996       Bill Clinton  Democratic      47400125    win  49.296938   
152  2000     George W. Bush  Republican      50456002    win  47.974666   
157  2004     George W. Bush  Republican      62040610    win  50.771824   
162  2008       Barack Obama  Democratic      69498516    win  53.023510   
168  2012       Barack Obama  Democratic      65915795    win  51.258484   
173  2016       Donald Trump  Republican      62984828    win  46.407862   
178  2020       Joseph Biden  Democratic      81268924    win  51.311515   

     Result_bool  
131         True  
133         True  
135         True  
140         True  
144         True  
152         True  
157         True  
162         True  
168         True  
173         True  
178         True

Above, we've assigned the result of the logical operator to a new variable called `iswin`. This is uncommon. Usually, the series is created and used on the same line. Such code is a little tricky to read at first, but you'll get used to it quickly.

Show all 'win' results between 1980 to 2000

In [135]:
#Answer Here
iswin.iloc[0:6]


Year          Candidate       Party  Popular vote Result          %  \
131  1980      Ronald Reagan  Republican      43903230    win  50.897944   
133  1984      Ronald Reagan  Republican      54455472    win  59.023326   
135  1988  George H. W. Bush  Republican      48886597    win  53.518845   
140  1992       Bill Clinton  Democratic      44909806    win  43.118485   
144  1996       Bill Clinton  Democratic      47400125    win  49.296938   
152  2000     George W. Bush  Republican      50456002    win  47.974666   

     Result_bool  
131         True  
133         True  
135         True  
140         True  
144         True  
152         True

Show all 'Loss' results of Independent party

In [136]:
# Answer Here
# Filtering for winners since 1980
loser_indep = elections[(elections['Party'] == 'Independent') & (elections['Result'] == 'loss')]
# Displaying the result
loser_indep

Year         Candidate        Party  Popular vote Result          %  \
121  1976   Eugene McCarthy  Independent        740460   loss   0.911649   
130  1980  John B. Anderson  Independent       5719850   loss   6.631143   
143  1992        Ross Perot  Independent      19743821   loss  18.956298   
161  2004       Ralph Nader  Independent        465151   loss   0.380663   
167  2008       Ralph Nader  Independent        739034   loss   0.563842   
174  2016     Evan McMullin  Independent        732273   loss   0.539546   

     Result_bool  
121        False  
130        False  
143        False  
161        False  
167        False  
174        False

We can select multiple criteria by creating multiple boolean Series and combining them using the `&` operator.

Show results of win with percentage less than 50%

In [144]:
# Answer Here
win_perc = elections[(elections['Result'] == 'win') & (elections['%'] < 50.0)]
win_perc

Year          Candidate                  Party  Popular vote Result  \
1    1824  John Quincy Adams  Democratic-Republican        113142    win   
16   1848     Zachary Taylor                   Whig       1360235    win   
20   1856     James Buchanan             Democratic       1835140    win   
23   1860    Abraham Lincoln             Republican       1855993    win   
33   1876   Rutherford Hayes             Republican       4034142    win   
36   1880     James Garfield             Republican       4453337    win   
39   1884   Grover Cleveland             Democratic       4914482    win   
43   1888  Benjamin Harrison             Republican       5443633    win   
47   1892   Grover Cleveland             Democratic       5553898    win   
70   1912     Woodrow Wilson             Democratic       6296284    win   
74   1916     Woodrow Wilson             Democratic       9126868    win   
100  1948       Harry Truman             Democratic      24179347    win   
117  1968      Richard Nixon             Republican      31783783    win   
140  1992       Bill Clinton             Democratic      44909806    win   
144  1996       Bill Clinton             Democratic      47400125    win   
152  2000     George W. Bush             Republican      50456002    win   
173  2016       Donald Trump             Republican      62984828    win   

             %  Result_bool  
1    42.789878         True  
16   47.309296         True  
20   45.306080         True  
23   39.699408         True  
33   48.471624         True  
36   48.369234         True  
39   48.884933         True  
43   47.858041         True  
47   46.121393         True  
70   41.933422         True  
74   49.367987         True  
100  49.601536         True  
117  43.565246         True  
140  43.118485         True  
144  49.296938         True  
152  47.974666         True  
173  46.407862         True

Show all 'win' results between 1980 to 2000

In [145]:
# Answer Here
iswin.iloc[0:6]


Year          Candidate       Party  Popular vote Result          %  \
131  1980      Ronald Reagan  Republican      43903230    win  50.897944   
133  1984      Ronald Reagan  Republican      54455472    win  59.023326   
135  1988  George H. W. Bush  Republican      48886597    win  53.518845   
140  1992       Bill Clinton  Democratic      44909806    win  43.118485   
144  1996       Bill Clinton  Democratic      47400125    win  49.296938   
152  2000     George W. Bush  Republican      50456002    win  47.974666   

     Result_bool  
131         True  
133         True  
135         True  
140         True  
144         True  
152         True

## Loc and iLoc

Show 5 enteries from start

In [148]:
elections.head()

Year          Candidate                  Party  Popular vote Result  \
0  1824     Andrew Jackson  Democratic-Republican        151271   loss   
1  1824  John Quincy Adams  Democratic-Republican        113142    win   
2  1828     Andrew Jackson             Democratic        642806    win   
3  1828  John Quincy Adams    National Republican        500897   loss   
4  1832     Andrew Jackson             Democratic        702735    win   

           %  Result_bool  
0  57.210122        False  
1  42.789878         True  
2  56.203927         True  
3  43.796073        False  
4  54.574789         True

In [150]:
# Answer Here
elections.loc[1:5]

Year          Candidate                  Party  Popular vote Result  \
1  1824  John Quincy Adams  Democratic-Republican        113142    win   
2  1828     Andrew Jackson             Democratic        642806    win   
3  1828  John Quincy Adams    National Republican        500897   loss   
4  1832     Andrew Jackson             Democratic        702735    win   
5  1832         Henry Clay    National Republican        484205   loss   

           %  Result_bool  
1  42.789878         True  
2  56.203927         True  
3  43.796073        False  
4  54.574789         True  
5  37.603628        False

You can provide `.loc` a list of row labels [0-5] and column labels ['Candidate','Party', 'Year'] as input to return a dataframe

In [162]:
#Answer Here
elections.loc[0:4, ['Candidate', 'Party', 'Year']]

Candidate                  Party  Year
0     Andrew Jackson  Democratic-Republican  1824
1  John Quincy Adams  Democratic-Republican  1824
2     Andrew Jackson             Democratic  1828
3  John Quincy Adams    National Republican  1828
4     Andrew Jackson             Democratic  1832

Loc also supports **slicing** (for all types, including numeric and string labels!). Note that the slicing for loc is **inclusive**, even for numeric slices.

Use Slicing on Rows and Columns

In [170]:
# Answer Here
elections.loc[1:5]

Year          Candidate                  Party  Popular vote Result  \
1  1824  John Quincy Adams  Democratic-Republican        113142    win   
2  1828     Andrew Jackson             Democratic        642806    win   
3  1828  John Quincy Adams    National Republican        500897   loss   
4  1832     Andrew Jackson             Democratic        702735    win   
5  1832         Henry Clay    National Republican        484205   loss   

           %  Result_bool  
1  42.789878         True  
2  56.203927         True  
3  43.796073        False  
4  54.574789         True  
5  37.603628        False

If we provide only a **single label** for the column argument, we get back a **Series**.

In [173]:
# Answer Here
elections.loc[:, 'Candidate']

0         Andrew Jackson
1      John Quincy Adams
2         Andrew Jackson
3      John Quincy Adams
4         Andrew Jackson
             ...        
177           Jill Stein
178         Joseph Biden
179         Donald Trump
180         Jo Jorgensen
181       Howard Hawkins
Name: Candidate, Length: 182, dtype: object

If we want a data frame instead and don't want to use to_frame, we can provide a **list** containing the column name.

In [175]:
# Answer Here
elections.loc[:, ['Candidate']]

Candidate
0       Andrew Jackson
1    John Quincy Adams
2       Andrew Jackson
3    John Quincy Adams
4       Andrew Jackson
..                 ...
177         Jill Stein
178       Joseph Biden
179       Donald Trump
180       Jo Jorgensen
181     Howard Hawkins

[182 rows x 1 columns]

If we give only one row but many column labels, we'll get back a **Series** corresponding to a row of the table. This new Series has a neat index, where **each entry is the name of the column** that the data came from.

In [185]:
# Answer Here
elections.loc[0, elections.columns[0:7]]

Year                             1824
Candidate              Andrew Jackson
Party           Democratic-Republican
Popular vote                   151271
Result                           loss
%                           57.210122
Result_bool                     False
Name: 0, dtype: object

If we omit the column argument altogether, the **default behavior is to retrieve all columns**.

In [183]:
# Answer Here
elections.loc[0]

Year                             1824
Candidate              Andrew Jackson
Party           Democratic-Republican
Popular vote                   151271
Result                           loss
%                           57.210122
Result_bool                     False
Name: 0, dtype: object

Specify Rows and Columns as List to retrive specific enteries

In [186]:
# Answer Here
rows=[1,2,3]
columns=[1,2,3]
elections.iloc[rows,columns]

Candidate                  Party  Popular vote
1  John Quincy Adams  Democratic-Republican        113142
2     Andrew Jackson             Democratic        642806
3  John Quincy Adams    National Republican        500897

Boolean Series are also boolean arrays, so we can use the Boolean Array Selection from earlier using loc as well.

In [194]:
# Answer Here
# Creating a Boolean Series
boolean_series = elections['Result'] == 'win'
winning_elections = elections.loc[boolean_series]
winning_elections.head()


Year               Candidate                  Party  Popular vote Result  \
1   1824       John Quincy Adams  Democratic-Republican        113142    win   
2   1828          Andrew Jackson             Democratic        642806    win   
4   1832          Andrew Jackson             Democratic        702735    win   
8   1836        Martin Van Buren             Democratic        763291    win   
11  1840  William Henry Harrison                   Whig       1275583    win   

            %  Result_bool  
1   42.789878         True  
2   56.203927         True  
4   54.574789         True  
8   52.272472         True  
11  53.051213         True

## String-labeled Rows

Let's do a quick example using data with string-labeled rows instead of integer labeled rows, just to make sure we're really understanding loc.

Use mottos.csv file

In [213]:
mottos=pd.read_csv('./data/mottos.csv',index_col='State')
mottos.head()

Motto                 Translation  \
State                                                                   
Alabama     Audemus jura nostra defendere  We dare defend our rights!   
Alaska                North to the future                           —   
Arizona                        Ditat Deus                God enriches   
Arkansas                   Regnat populus             The people rule   
California                Eureka (Εὕρηκα)             I have found it   

           Language Date Adopted  
State                             
Alabama       Latin         1923  
Alaska      English         1967  
Arizona       Latin         1863  
Arkansas      Latin         1907  
California    Greek         1849

In [214]:
# Answer Here
mottos.loc['Arizona', ['Motto', 'Language']]

Motto       Ditat Deus
Language         Latin
Name: Arizona, dtype: object

Extract slice, can be specified using slice notation, even if the rows have string labels instead of integer labels.

### iloc

loc's cousin iloc is very similar, but is used to access based on numerical position instead of label. For example, to access to the top 3 rows and top 3 columns of a table, we can use [0:3, 0:3]. 'iloc' slicing is **exclusive**, just like standard Python slicing of numerical values.

Use iloc to extract first 3 rows and columns from elections DataFrame

In [216]:
#Answer Here
elections.iloc[0:3,0:3]

Year          Candidate                  Party
0  1824     Andrew Jackson  Democratic-Republican
1  1824  John Quincy Adams  Democratic-Republican
2  1828     Andrew Jackson             Democratic

We will use both `loc` and `iloc` in the course. `loc` is generally preferred for a number of reasons, for example:

1. It is harder to make mistakes since you have to literally write out what you want to get.
2. Code is easier to read, because the reader doesn't have to know e.g., what column #17 represents.
3. It is robust against permutations of the data, e.g. the social security administration switches the order of two columns.

However, iloc is sometimes more convenient. We'll provide examples of when iloc is the superior choice.

## Handy Properties and Utility Functions for Series and DataFrames

The head, shape, size, and describe methods can be used to quickly get a good sense of the data we're working with. For example:

In [217]:
mottos = pd.read_csv("./data/mottos.csv")

In [225]:
# Answer Here
mottos.head()

State                          Motto                 Translation  \
0     Alabama  Audemus jura nostra defendere  We dare defend our rights!   
1      Alaska            North to the future                           —   
2     Arizona                     Ditat Deus                God enriches   
3    Arkansas                 Regnat populus             The people rule   
4  California                Eureka (Εὕρηκα)             I have found it   

  Language Date Adopted  
0    Latin         1923  
1  English         1967  
2    Latin         1863  
3    Latin         1907  
4    Greek         1849

Size of DataFrame

In [223]:
# Answer Here
mottos.size

250

The fact that the size is 250 means our data file is relatively small, with only 250 total entries.

Shape of DataFrame

In [224]:
# Answer Here
print(f'The mottos data has {mottos.shape[0]} rows and {mottos.shape[1]}')

The mottos data has 50 rows and 5


Use describe function and extract the meaningful information from DataFrame

In [226]:
# Answer Here
mottos.describe()

State                          Motto Translation Language  \
count        50                             50          49       50   
unique       50                             50          30        8   
top     Alabama  Audemus jura nostra defendere           —    Latin   
freq          1                              1          20       23   

       Date Adopted  
count            50  
unique           47  
top            1893  
freq              2

Above, we see a quick summary of all the data. For example, the most common language for mottos is Latin, which covers 23 different states. Does anything else seem surprising?

We can get a direct reference to the index using .index.

In [227]:
# Answer Here
mottos.index

RangeIndex(start=0, stop=50, step=1)

In [228]:
mottos.head(2)

State                          Motto                 Translation  \
0  Alabama  Audemus jura nostra defendere  We dare defend our rights!   
1   Alaska            North to the future                           —   

  Language Date Adopted  
0    Latin         1923  
1  English         1967

It turns out the columns also have an Index. We can access this index by using `.columns`.

In [229]:
# Answer Here
mottos.columns

Index(['State', 'Motto', 'Translation', 'Language', 'Date Adopted'], dtype='object')

## Sorting and Value Counts

There are also a ton of useful utility methods we can use with Data Frames and Series. For example, we can create a copy of a data frame sorted by a specific column using `sort_values`.

In [230]:
# Answer Here
mottos.sort_values(by='State')

State                                              Motto  \
0          Alabama                      Audemus jura nostra defendere   
1           Alaska                                North to the future   
2          Arizona                                         Ditat Deus   
3         Arkansas                                     Regnat populus   
4       California                                    Eureka (Εὕρηκα)   
5         Colorado                                    Nil sine numine   
6      Connecticut                            Qui transtulit sustinet   
7         Delaware                           Liberty and Independence   
8          Florida                                    In God We Trust   
9          Georgia                        Wisdom, Justice, Moderation   
10          Hawaii                  Ua mau ke ea o ka ʻāina i ka pono   
11           Idaho                                      Esto perpetua   
12        Illinois                  State sovereignty, national union   
13         Indiana                          The Crossroads of America   
14            Iowa  Our liberties we prize and our rights we will ...   
15          Kansas                                Ad astra per aspera   
16        Kentucky                               Deo gratiam habeamus   
17       Louisiana                         Union, justice, confidence   
18           Maine                                             Dirigo   
19        Maryland                       Fatti maschii, parole femine   
20   Massachusetts          Ense petit placidam sub libertate quietem   
21        Michigan          Si quaeris peninsulam amoenam circumspice   
22       Minnesota                                   L'étoile du Nord   
23     Mississippi                                   Virtute et armis   
24        Missouri                      Salus populi suprema lex esto   
25         Montana                                        Oro y plata   
26        Nebraska                            Equality before the law   
27          Nevada                                All For Our Country   
28   New Hampshire                                   Live Free or Die   
29      New Jersey                             Liberty and prosperity   
30      New Mexico                                      Crescit eundo   
31        New York                                          Excelsior   
32  North Carolina                                   Esse quam videri   
33    North Dakota                    \nSerit ut alteri saeclo prosit   
34            Ohio                  With God, all things are possible   
35        Oklahoma                                 Labor omnia vincit   
36          Oregon                                Alis volat propriis   
37    Pennsylvania                  Virtue, liberty, and independence   
38    Rhode Island                                               Hope   
39  South Carolina          Dum spiro spero \nAnimis opibusque parati   
40    South Dakota                          Under God the people rule   
41       Tennessee                           Agriculture and Commerce   
42           Texas                                         Friendship   
43            Utah                                           Industry   
44         Vermont                       Stella quarta decima fulgeat   
45        Virginia                                Sic semper tyrannis   
46      Washington                                      Al-ki or Alki   
47   West Virginia                              Montani semper liberi   
48       Wisconsin                                            Forward   
49         Wyoming                                       Equal Rights   

                                          Translation        Language  \
0                          We dare defend our rights!           Latin   
1                                                   —         English   
2                                        God enriches           Latin 

As mentioned before, all Data Frame methods return a copy and do **not** modify the original data structure, unless you set inplace to True.

If we want to sort in reverse order, we can set `ascending=False`.

In [231]:
elections.sort_values('%', ascending=False)

Year           Candidate         Party  Popular vote Result          %  \
114  1964      Lyndon Johnson    Democratic      43127041    win  61.344703   
91   1936  Franklin Roosevelt    Democratic      27752648    win  60.978107   
120  1972       Richard Nixon    Republican      47168710    win  60.907806   
79   1920      Warren Harding    Republican      16144093    win  60.574501   
133  1984       Ronald Reagan    Republican      54455472    win  59.023326   
..    ...                 ...           ...           ...    ...        ...   
165  2008    Cynthia McKinney         Green        161797   loss   0.123442   
148  1996        John Hagelin   Natural Law        113670   loss   0.118219   
160  2004    Michael Peroutka  Constitution        143630   loss   0.117542   
141  1992            Bo Gritz      Populist        106152   loss   0.101918   
156  2004          David Cobb         Green        119859   loss   0.098088   

     Result_bool  
114         True  
91          True  
120         True  
79          True  
133         True  
..           ...  
165        False  
148        False  
160        False  
141        False  
156        False  

[182 rows x 7 columns]

We can also use `sort_values` on Series objects.

In [232]:
mottos['Language'].sort_values().head(50)

46    Chinook Jargon
49           English
29           English
28           English
27           English
26           English
48           English
37           English
38           English
40           English
17           English
34           English
42           English
14           English
41           English
12           English
1            English
13           English
8            English
7            English
9            English
43           English
22            French
4              Greek
10          Hawaiian
19           Italian
39             Latin
44             Latin
36             Latin
45             Latin
47             Latin
35             Latin
33             Latin
0              Latin
31             Latin
30             Latin
23             Latin
21             Latin
20             Latin
18             Latin
16             Latin
15             Latin
11             Latin
6              Latin
5              Latin
3              Latin
2              Latin
32           

For Series, the `value_counts` method is often quite handy.

In [233]:
mottos['Language'].value_counts()

Language
Latin             23
English           21
Greek              1
Hawaiian           1
Italian            1
French             1
Spanish            1
Chinook Jargon     1
Name: count, dtype: int64

Also commonly used is the `unique` method, which returns **all unique values** as a numpy array.

In [234]:
mottos['Language'].unique()

array(['Latin', 'English', 'Greek', 'Hawaiian', 'Italian', 'French',
       'Spanish', 'Chinook Jargon'], dtype=object)

In [235]:
def fiba(n):
    if n < 2:
        return n
    else:
        return fiba(n-1) + fiba(n-2)



fiba(5)

5

# Thank you!